micro-rdk/src/common/data_manager.rs

```diff
@@ -610,6 +635,117 @@ where
+                .call_method(self.robot_start_time)
+                .inspect_err(|err| {
+                    log::error!(
+                        "failed to collect readings for {:?}: {:?}",
+                        collector_key,
+                        err
+                    )
+                });
+            if let Ok(mut readings) = readings {
+                match time_correct_reading(self.robot_start_time, &mut readings) {
+                    Ok(_) => {
+                        let upload_request = DataCaptureUploadRequest {
+                            metadata: Some(UploadMetadata {
+                                part_id: self.part_id.clone(),
+                                component_type: collector_key.component_type.clone(),
+                                r#type: DataType::TabularSensor.into(),
+                                component_name: collector_key.r_name.clone(),
+                                method_name: collector_key.method.to_string(),
+                                ..Default::default()
+                            }),
+                            sensor_contents: vec![readings],
+                        };
+                        match app_client.upload_data(upload_request).await {
+                            Ok(_) => {
+                                #[cfg(feature = "data-upload-hook-unstable")]
+                                unsafe {
+                                    micro_rdk_data_manager_post_upload_hook();
+                                }
+                            }
+                            Err(err) => {
+                                log::error!("error uploading data, failed to upload {:?} readings on this attempt", collectors_len - idx);
+                                return Err(err);
+                            }
+                        };
+                    }
+                    Err(err) => {
+                        log::error!(
+                            "unable to time correct reading for collector {:?}, {:?}",
+                            collector_key,
+                            err
+                        )
+                    }
+                };
+            }
+        }
+        let _ = send_system_event(SystemEvent::DeepSleep(Some(self.sync_interval)), true).await;
+        Ok(())
+    }
+
+    pub fn from_robot_and_config(
+        robot: &LocalRobot,
+        cfg: &RobotConfig,
+        robot_start_time: Instant,
+    ) -> Result<Self, DataManagerError> {
+        if let Some(cfg) = get_data_service_config(cfg)? {
+            let (collectors, sync_interval, part_id) =
+                get_collectors_and_settings_from_robot_and_config(robot, &cfg)?;
+            match sync_interval {
+                None => Err(DataManagerError::MissingSyncInterval),
+                Some(sync_interval) => {
+                    if collectors.iter().any(|c| {
+                        (c.time_interval().as_secs_f32() - sync_interval.as_secs_f32()).abs()
```

Comment:
not sure what we are checking for here, can you provide more background?

Reply:
will put a comment detailing this, but just to answer the question; we discussed wanting to temporarily validate that every collection frequency was equivalent to the sync frequency when using this mode. However due to variances in float division we cannot merely check "==", so we need to instead validate that the discrepancy in seconds is lower than a tolerated threshold